### ISSUE TITLE
percentile calculation issue?

### ISSUE THREAD

**Issue by fsolleza** [numpy/numpy#8821]
If I do the following:

    >>> arr = np.arange(1, 101, 1)
    >>> arr.shape
    (100,)
    >>> np.percentile(arr, 10, interpolation="higher")
    11

Shouldn't the result = 10? My suspicion is that in numpy.lib.function_base._percentile, the index of the percentile is calculated as:

    Nx = ap.shape[axis]
    indices = q * (Nx - 1)

The problem here is that in the example I gave, the base is now 99, instead of 100 and `indices` is now 9.9. After this,

    elif interpolation == 'higher':
        indices = ceil(indices).astype(intp)

yields an index of 10.

Instead, I think `indices` should be calculated as:

    Nx = ap.shape[axis]
    indices = (q * Nx)

then 

    elif interpolation == 'higher':
        indices = ceil(indices).astype(intp) - 1

Which would yield the correct result. Note that I think this issue affects all methods of calculating percentiles, whether interpolated or not.

I _think_ what I said is right, but I'm happy to be corrected...

I'm also happy to make the contribution (and learn the manner in which contributions are made).

**Comment by eric-wieser** [comment]
A simpler case - what do you think `np.percentile([0, 1], 50, interpolation="higher")` should give? Your rule would produce `0`, I think

**Comment by fsolleza** [comment]
I'm not so sure about that. I'm still working on a few things to account for all the other interpolations but isn't `0` w/in the 50th percentile, and `1` w/in the 100th percentile of two data points whose values are 0 and 1? Percentiles are based on the order/rank of values rather than the values themselves. If so I would expect `np.percentile([..any two values here..], 50, interpolate="higher")` to return the lower of the two, and 1, to return the higher of the two. Similarly, any of the other methods.

**Comment by eric-wieser** [comment]
Yes, percentiles only care about the order (ignoring interpolation), 0 and 1 were just examples. Better to discuss would be `np.percentile(['first', 'second'])`, I guess

But I would say that every percentile from 0%-

**Comment by fsolleza** [comment]
Yep - I agree - my recommended change is not right - ergo I'm spending a bit more time trying to put some edits together that are more "right" so to speak. Although wrong is wrong regardless of how subtle. I'm hoping to get something put together w/in the weekend.

**Comment by eric-wieser** [comment]
Actually, your change looks like an implementation of interpolation="lower"  to me

**Comment by fsolleza** [comment]
Yeah I'd say, I'll work on another recommended change. The attempt there was to account for the indices calculated based on the true length of the array, as opposed to the length - 1 which was the origin error in the first place - since the denominator of the percentile would now be off by 1.

A simpler example would be a `np.percentile(np.arange(1,11,1), 10, interpolation="higher")` yielding `2` instead of `1` for two reasons: first: `indices = .1*9` gets ceiled to `1`, and second it doesn't account that the array is zero indexed.

Really, 10th percentile of a set of 10 is the first element is `indices = .1*10 = 1`. If the array is zero indexed, then the position would be `indices-1 = 0` which would stay as zero regardless of `ceil` or `floor` or any other method. Though, again, I maybe wrong about this understanding.

**Comment by fsolleza** [comment]
To add, given the multitude of ways to calculate percentiles, I've got to also qualify my comments by saying, maybe this is all a wild goose chase and the numpy calculations are probably right....

**Comment by jlijffijt** [comment]
I think the current implementation of 'higher' is wrong, but I do not have enough knowledge to fix the code.

Example with 39 entries:
`data = [ 0.72,  0.85,  1.09,  1.24,  1.37,  1.4,   1.47,  1.51,  1.58,  1.63,  1.64,  1.64,  1.67,  1.68,  1.69,  1.7,   1.74,  1.75,  1.75,  1.79,  1.79,  1.82,  1.85,  1.86,  1.88,  1.9,   1.92,  1.93,  1.97,  2.03,  2.08,  2.09,  2.11,  2.17,  2.28,  2.31,  2.37,  2.46,  2.55,]`

`numpy.percentile(data,25,interpolation='higher')` gives 1.64, which is the 11th data point. Since 0.25*39 = 9.75, I don't understand why rounding that to 11 makes sense. This behaviour is definitely not apparent from the documentation.

In contrast, `numpy.percentile(data,75,interpolation='higher')` gives 2.03 which is indeed the ceil(0.75*39) = ceil(29.25) = 30th data point. I think there is simply a bug in the implementation.

**Comment by fsolleza** [comment]
Ill try to find the time to work on this bug this week. Thanks!

**Comment by jlijffijt** [comment]
Sorry I have been too quick about this. It actually is a design choice, not a bug.

There are many ways to define percentiles, see for example the [R implementation of quantile()](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/quantile.html). The current implementation of percentile() corresponds to defining an empirical distribution where the first element sits at 0%, second at 1/(n-1), third at 2/(n-1) up to the nth at (n-1)/(n-1). I guess this has been derived from the design perspective that the median should sit at 0.5, instead of at (n+1)/(2n) as in the definition that I consider more intuitive in general.

My advice would be to update the documentation accordingly, instead of changing the percentile() function. Because who knows what could break and there is no universal truth here.

Anyway, the way we teach the empirical distribution function to our engineering students is that the first element sits at 1/n, second at 2/n, etc. This is also what Matlab does, and the default choice in R, for example.

The definitions are quite similar and it is easy to convert from one scheme to another: just insert the smallest element into your data array again.

I found out about this behaviour through the following experiment.
`import numpy

z = range(1,3)

for i in range(0,101):
    print(numpy.percentile(z,i,interpolation='higher'))
`

All values except the first are 2. To get the behaviour fsolleza and I were looking for, just define `z = [1, 1, 2]` instead.

To summarize: the documentation should be changed to clearly mark the non-standard choice of how the empirical distribution function is defined. Otherwise, the implementation can stay as-is. Perhaps an extra parameter could be introduced to allow people to work with the more intuitive notion.

**Comment by fsolleza** [comment]
I think that's where we fell at the last time. I do agree changing the implementation might be a little too dangerous and was considering alternatives before getting too busy. However, I also dont think that having the user work around the API is good either (someone is eventually going to make the same mistake).

**Comment by seberg** [comment]
Its tricky stuff, documenting it better would be very nice. Doubling checking all the methods, and possibly introducing new options, etc. may also make sense. If someone who is a specialist in the area thinks one `interpolation` option is wrong/weird, we could even consider deprecating and replacing it in the very long run.

**Comment by jlijffijt** [comment]
@seberg `interpolation` appears to work completely as expected, given the nonstandard elementary behaviour of the function. However, percentile() generally works unlike expected. For example, it is unlike [the definition in Wikipedia](https://en.wikipedia.org/wiki/Percentile).

**Comment by jlijffijt** [comment]
I consulted with a few of my colleagues. I asked them to draw me the empirical cumulative distribution function for some data. They (PhD students and more senior researchers in Data Mining) all gave the same answer. Based on this, my advice is that the most sensible default behaviour would be that the internal distribution function values used in `percentile()` are exactly the inverse of [`scipy.percentileofscore(..., kind='weak')`](https://github.com/scipy/scipy/blob/v0.19.0/scipy/stats/stats.py#L1636-L1729).

**Comment by eric-wieser** [comment]
Can you include their answer and the data in a comment here?

**Comment by seberg** [comment]
That is nice, though I would prefer statisticians ;). But that probably means checking again what R suggests (not necessarily their default). Just that I am not sure if the expected default is necessarily the best default.

**Comment by jlijffijt** [comment]
You should prefer statisticians ;-)

I am not certain whether the default behaviour of R and Matlab is exactly the same; they both use some real form of interpolation (e.g., not like `interpolation = 'higher'`) and that may be slightly different. I will make comparison with test cases next week and post it here.

**Comment by seberg** [comment]
I am pretty certain it is not the same, the last time I thought about these things, I gave up at some point, checked that some of the stuff we did matched Rs and hoped for the best....

**Comment by mattip** [comment]
We have improved the [documentation](http://www.numpy.org/devdocs/reference/generated/numpy.percentile.html) with a visualization of the different interpolation methods at the bottom, does that help clear up the issue?

**Comment by eric-wieser** [comment]
I think #10736 probably covers the lingering concerns

**Comment by seberg** [comment]
Yeah, I agree with Eric, there are some serious concerns here, but this can be closed in favor of that issue which includes some more information about what exactly is the problem.
